feat: add create tax component

OussEmaDevCode opened the pull request:
## Description
Added the create tax component, connected it via routing, designed its interface and connected it to the API.
## Related issues and discussion
#198 

## Screenshots, if any
![Capture d’écran (236)](https://user-images.githubusercontent.com/38397893/72211667-8f94f500-34cf-11ea-9c9b-ae980bd01555.png)


## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

OussEmaDevCode commented:
@abhaychawla @Jov03 Hi, please have a look !

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.html:
```diff
@@ -0,0 +1,71 @@
+            </mat-error>
+            <mat-error *ngIf="taxComponentForm.controls.percentage.hasError('min')">
+              Percentage <strong>cannot</strong> be smaller than zero
+            </mat-error>
```
The percentage should be between 1 and 100. Make appropriate changes.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.html:
```diff
@@ -0,0 +1,71 @@
+              Percentage <strong>cannot</strong> be smaller than zero
+            </mat-error>
+          </mat-form-field>
+
```
Please add the fields for Debit Account Type and Debit Account as well: https://github.com/openMF/community-app/blob/develop/app/views/products/tax/CreateTaxComponent.html 
Ensure they are not hidden in the web app.

Make appropriate changes to show Credit Account Field as mandatory if Credit Account Type is selected. Similar case should be handled for Debit Account as well.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.html:
```diff
@@ -0,0 +1,71 @@
+          <mat-error *ngIf="taxComponentForm.controls.startDate.hasError('required')">
+            Start date is <strong>required</strong>
+          </mat-error>
+        </mat-form-field>
```
This field should also cover the entire length. The `div` is incorrectly placed. Please check.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,115 @@
+  creditAccountTypeData: any;
+  /** Credit Account data. */
+  creditAccountData: any;
+  /**
```
```suggestion

  /**
```

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,115 @@
+  /**
+   * Retrieves the tax Component template data from `resolve`.
+   * @param {FormBuilder} formBuilder Form Builder.
+   * @param {OrganizationService} productsService Products Service.
```
Wrong parameter.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.html:
```diff
@@ -0,0 +1,71 @@
+
+          <mat-form-field>
+            <mat-label>Credit Account Type</mat-label>
+            <mat-select formControlName="creditAccountType" (selectionChange)="populateCreditAccount()">
```
```suggestion
            <mat-select formControlName="creditAccountType" (selectionChange)="populateCreditAccount()">
```
Subscribe to form control's value changes instead of using the selectionChange event. Have a look at other components for reference.

abhaychawla commented on src/app/products/manage-tax-components/manage-tax-components.component.html:
```diff
@@ -1,40 +1,40 @@
       Create Tax Component
     </button>
   </div>
-  
```
Remove changes to spacing which is not a part of this PR.

OussEmaDevCode commented:
@abhaychawla All required changes have been done, I've also resolved merge conflicts, updated the PR message with the new UI and changed the redirection to view tax component.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.html:
```diff
@@ -0,0 +1,98 @@
+            <mat-error *ngIf="taxComponentForm.controls.debitAcountId.hasError('required')">
+              Debit account is <strong>required</strong>
+            </mat-error>
+          </mat-form-field>
```
Similar to other components in the web app (for example create user), lets keep the control conditionally hidden and add/remove it accordingly from code instead of update. This way we can also ensure that `required` attribute is added to this field.

Do similar implementation for Credit Account.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,148 @@
+  /** Minimum start date allowed. */
+  minDate = new Date();
+  /** Maximum start date allowed. */
+  maxDate = new Date(2025, 6, 22);
```
Any specific reason to keep the max date as this value? If no, please keep it the same as in other components (in general).

OussEmaDevCode replied:
That's the value I found in the current implementation in the community app [here](https://github.com/openMF/community-app/blob/develop/app/scripts/controllers/product/tax/CreateTaxComponentController.js)

abhaychawla replied:
Okay let us update this, so it is not hardcoded:
```suggestion
  maxDate = new Date(new Date().setFullYear(new Date().getFullYear() + 10));
```

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,148 @@
+    this.creditAccountTypeData = this.debitAccountTypeData = this.taxComponentTemplateData.glAccountTypeOptions;
+    this.taxComponentForm = this.formBuilder.group({
+      'name': ['', Validators.required],
+      'percentage': ['', [Validators.required, Validators.min(0), Validators.max(100)]],
```
```suggestion
      'percentage': ['', [Validators.required, Validators.min(1), Validators.max(100)]],
```

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,148 @@
+  ngOnInit() {
+    this.createTaxComponentForm();
+    this.setCreditAccountsForm();
+    this.setDebitAccountsForm();
```
Add a single function to set conditional controls. Add/Remove control conditional instead of update for this case.

For example, https://github.com/openMF/web-app/blob/master/src/app/users/create-user/create-user.component.ts

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,148 @@
+        accountsData = glAccountOptionsData.expenseAccountOptions;
+        return accountsData;
+      default:
+        break;
```
Not required.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,148 @@
+    switch (accountTypeId) {
+      case 1:
+        accountsData = glAccountOptionsData.assetAccountOptions;
+        return accountsData;
```
No need for extra variable, just return the data directly for all cases.

OussEmaDevCode commented:
@abhaychawla Done !

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.html:
```diff
@@ -0,0 +1,96 @@
+      </mat-card-content>
+
+      <mat-card-actions fxLayout="row" fxLayout.xs="column" fxLayoutAlign="center" fxLayoutGap="5px">
+        <button mat-raised-button [routerLink]="['../']">Cancel</button>
```
```suggestion
        <button type="button" mat-raised-button [routerLink]="['../']">Cancel</button>
```
If you don't add this, cancel button will also submit the form.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.html:
```diff
@@ -0,0 +1,96 @@
+
+          <mat-form-field *ngIf="taxComponentForm.controls.creditAccountType.value">
+            <mat-label>Credit Account</mat-label>
+            <mat-select formControlName="creditAcountId">
```
```suggestion
            <mat-select required formControlName="creditAcountId">
```
Add required attribute.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.html:
```diff
@@ -0,0 +1,96 @@
+
+          <mat-form-field *ngIf="taxComponentForm.controls.debitAccountType.value">
+            <mat-label>Debit Account</mat-label>
+            <mat-select formControlName="debitAcountId">
```
```suggestion
            <mat-select required formControlName="debitAcountId">
```
Add required attribute.

abhaychawla commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.html:
```diff
@@ -0,0 +1,96 @@
+            </mat-error>
+            <mat-error *ngIf="taxComponentForm.controls.percentage.hasError('min') ||
+            taxComponentForm.controls.percentage.hasError('max')">
+              Percentage <strong>should</strong> be between one and a hundred
```
```suggestion
              Percentage <strong>should</strong> be between 1 and 100
```

punwai replied:
@abhaychawla I was just wondering, wouldn't it be better if the percentage have to be larger than 0 and at most 100, so that the user will be able to enter decimal tax values such as 0.4% etc.

punwai replied:
If that is the case, he can simply reuse the regex from the create-tax file to do this too, as angular doesn't provide a larger than validator.

abhaychawla replied:
Alright, @punwai I checked. @OussEmaDevCode it seems decimal values are allowed by the API. Let us use a `regex` pattern of decimal values greater than 0 instead of min and max validators and update the error message as he specified above.

punwai replied:
@OussEmaDevCode Could I suggest that for your mental health, you use the same regex to validate for percentages larger than 0, but then still use a max to validate for it to be at most 100. It will be a very complicated regex to deal with both inequalities.

OussEmaDevCode replied:
Good suggestion !!

punwai commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,129 @@
+  setConditionalControls() {
+    this.taxComponentForm.get('debitAccountType').valueChanges.subscribe(debitAccountTypeId => {
+      this.debitAccountData = this.getAccountsData(debitAccountTypeId);
+      this.taxComponentForm.addControl('debitAcountId', new FormControl('', Validators.required));
```
```suggestion
      this.taxComponentForm.addControl('debitAccountId', new FormControl('', Validators.required));
```

OussEmaDevCode replied:
@punwai That's not a typo that's the current fineract API implementation

punwai replied:
@OussEmaDevCode My mistakes

abhaychawla replied:
This is not an issue, it is the API format. The typo needs to be corrected at the server end first.

punwai commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,129 @@
+    });
+    this.taxComponentForm.get('creditAccountType').valueChanges.subscribe(creditAccountTypeId => {
+      this.creditAccountData = this.getAccountsData(creditAccountTypeId);
+      this.taxComponentForm.addControl('creditAcountId', new FormControl('', Validators.required));
```
```suggestion
      this.taxComponentForm.addControl('creditAccountId', new FormControl('', Validators.required));
```

abhaychawla replied:
This is not an issue, it is the API format. The typo needs to be corrected at the server end first.

OussEmaDevCode replied:
@punwai Again, that's not a typo that's the current fineract API implementation

punwai commented on src/app/products/manage-tax-components/create-tax-component/create-tax-component.component.ts:
```diff
@@ -0,0 +1,129 @@
+   * @returns {any} Accounts data
+   */
+  getAccountsData(accountTypeId: number) {
+    const glAccountOptionsData = this.taxComponentTemplateData.glAccountOptions;
```
Can't this be defined at the top of the file?

OussEmaDevCode replied:
This variable only has one use case in that function, it'll be useless to make it a global variable.

abhaychawla replied:
No. We can remove this variable though as its extra and use `taxComponentTemplateData` directly.

OussEmaDevCode replied:
okay

OussEmaDevCode commented:
@abhaychawla Done !

luckyman20 commented:
@OussEmaDevCode Please resolve merge conflicts

OussEmaDevCode commented:
@luckyman20 Done !
